### PR TITLE
fixed Python 3 keys() usage

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY
Searched the repo for usages of `dict.keys()` in a `list` context, something that worked in python2 but not in python3. Wrapped those calls with `list()`.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins/cache/redis.py
plugins/callback/diy.py
plugins/callback/selective.py
plugins/lookup/chef_databag.py
plugins/module_utils/net_tools/nios/api.py
plugins/module_utils/utm_utils.py
plugins/modules/cloud/lxc/lxc_container.py
plugins/modules/cloud/lxd/lxd_container.py
plugins/modules/cloud/oneandone/oneandone_monitoring_policy.py
plugins/modules/cloud/oracle/oci_vcn.py
plugins/modules/cloud/spotinst/spotinst_aws_elastigroup.py
plugins/modules/monitoring/sensu/sensu_check.py
plugins/modules/packaging/os/redhat_subscription.py
plugins/modules/remote_management/redfish/idrac_redfish_command.py
plugins/modules/remote_management/redfish/idrac_redfish_config.py
plugins/modules/remote_management/redfish/idrac_redfish_info.py
plugins/modules/remote_management/redfish/redfish_command.py
plugins/modules/remote_management/redfish/redfish_config.py
plugins/modules/system/vdo.py
scripts/inventory/nsot.py

### CODE REVIEW

**Comment by felixfontein** [comment]
It still works with Python 3 - but as opposed to Python 2 it returns an iterator and not a list. So it depends on the context.

File: plugins/cache/redis.py [hunk 1]
```diff
@@ -217,13 +217,13 @@ def delete(self, key):
...
     def flush(self):
-        for key in self.keys():
+        for key in list(self.keys()):
             self.delete(key)
 
     def copy(self):
         # TODO: there is probably a better way to do this in redis
         ret = dict()
-        for key in self.keys():
+        for key in list(self.keys()):
```
Comment (felixfontein) [inline plugins/cache/redis.py:226]:
```suggestion
        for key in self.keys():
```
This only makes it less efficient both on Python 2 and Python 3.

File: plugins/modules/cloud/spotinst/spotinst_aws_elastigroup.py [hunk 2]
```diff
@@ -1305,10 +1305,12 @@ def expand_tags(eg_launchspec, tags):
...
-            if tag.keys():
-                eg_tag.tag_key = tag.keys()[0]
-            if tag.values():
-                eg_tag.tag_value = tag.values()[0]
+            key_list = list(tag.keys())
+            val_list = list(tag.values())
+            if key_list:
+                eg_tag.tag_key = key_list[0]
+            if val_list:
+                eg_tag.tag_value = val_list[0]
```
Comment (felixfontein) [inline plugins/modules/cloud/spotinst/spotinst_aws_elastigroup.py:1313]:
```suggestion
            if tag:
                eg_tag.tag_key, eg_tag.tag_value = list(tag.items())[0]
```

File: plugins/modules/monitoring/sensu/sensu_check.py [hunk 3]
```diff
@@ -272,7 +272,7 @@ def sensu_check(module, path, name, state='present', backup=False):
                     changed = True
                     reasons.append('`custom param {opt}\' was added'.format(opt=k))
                 check[k] = v
-            simple_opts += custom_params.keys()
+            simple_opts += list(custom_params.keys())
```
Comment (felixfontein) [inline plugins/modules/monitoring/sensu/sensu_check.py:275]:
```suggestion
            simple_opts += custom_params.keys()
```

**Comment by russoz** [comment]
> It still works with Python 3 - but as opposed to Python 2 it returns an iterator and not a list. So it depends on the context.

yes, of course:
```
$ git grep -nE '[a-zA-Z0-9_\.]+\.keys\(\)' | grep -vE '(list\(|sorted\(|join\(|set\(|len\(|in )[^\(\)]+\.keys\(\)'
```

**Review by felixfontein — COMMENTED** [review]
I think all other cases are ok.

File: plugins/modules/monitoring/sensu/sensu_check.py (hunk 3 above)
Comment (russoz) [reply plugins/modules/monitoring/sensu/sensu_check.py:275]: Not sure that `list` + `dict_keys` work

Comment (russoz) [reply plugins/modules/monitoring/sensu/sensu_check.py:275]: In fact it doesn't:
```python
>>> a = dict(aa=1, ab=2, ac=3)
>>> [] + a.keys()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate list (not "dict_keys") to list
```

Comment (felixfontein) [reply plugins/modules/monitoring/sensu/sensu_check.py:275]: It does:
```
>>> a = [1]
>>> b = {2: 3}
>>> a += b.keys()
>>> a
[1, 2]
```

Comment (felixfontein) [reply plugins/modules/monitoring/sensu/sensu_check.py:275]: (That only works with `+=` though, not with `+`.)

Comment (felixfontein) [reply plugins/modules/monitoring/sensu/sensu_check.py:275]: Please note that you tested `+`, while the code has `+=`.

**Comment by russoz** [comment]
btw, I did not include memcached as it is being handled in #1857

File: plugins/modules/monitoring/sensu/sensu_check.py (hunk 3 above)
Comment (russoz) [reply plugins/modules/monitoring/sensu/sensu_check.py:275]: oooohhhhh shoot... why on earth do they behave differently? fixing...

**Comment by russoz** [comment]
ha, it seems we're also getting some ignore lines off with that :-) win-win

File: changelogs/fragments/1861-python3-keys.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+bugfixes:
+  - redis - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - memcached - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - diy - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - selective - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - chef_databag - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - api - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - utm_utils - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
```
Comment (felixfontein) [inline changelogs/fragments/1861-python3-keys.yml:8]:
```suggestion
  - redis cache plugin - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
  - memcached cache plugin - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
  - diy callback plugin - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
  - selective callback plugin - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
  - chef_databag lookup plugin - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
  - net_tools.nios.api module_utils - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
  - utm_utils module_utils - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
```

**Comment by mraineri** [comment]
Reviewed the Redfish modules; LGTM!

File: changelogs/fragments/1861-python3-keys.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  - idrac_redfish_command - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - idrac_redfish_config - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - idrac_redfish_info - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - redfish_command - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - redfish_config - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - vdo - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - nsot - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
```
Comment (russoz) [inline changelogs/fragments/1861-python3-keys.yml:22]:
```suggestion
  - nsot script - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
```

File: changelogs/fragments/1861-python3-keys.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  - idrac_redfish_command - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - idrac_redfish_config - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - idrac_redfish_info - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - redfish_command - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - redfish_config - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - vdo - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
+  - nsot script - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
```
Comment (felixfontein) [inline changelogs/fragments/1861-python3-keys.yml:22]:
```suggestion
  - nsot inventory script - wrapped usages of ``dict.keys()`` in ``list()`` for Python 3 compatibility (https://github.com/ansible-collections/community.general/pull/1861).
```

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
I will keep this open for a bit longer, in case more maintainers find time to take a look.

**Comment by felixfontein** [comment]
@russoz thanks for fixing this!
@mraineri thanks for reviewing!

**Pull request merged**